File: hellojava/src/test/java/com/fathzer/chess/MoveTest.java
package com.fathzer.chess;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.Test;

class MoveTest {
    @Test
    void testValidMove() {
        // Test valid move construction
        Move move = new Move(0, 16); // a1 to a3
        assertEquals(0, move.from());
        assertEquals(16, move.to());
    }

    @Test
    void testToString() {
        // Test standard move
        Move e2e4 = new Move("e2", "e4"); // e2 to e4
        assertEquals("e2e4", e2e4.toString());
    }
    
    @Test
    void testEquality() {
        // Test record's built-in equality
        Move move1 = new Move(0, 16);
        Move move2 = new Move(0, 16);
        Move move3 = new Move(16, 0);
        
        assertEquals(move1, move2);
        assertEquals(move1.hashCode(), move2.hashCode());
        assertNotEquals(move1, move3);
    }
}

File: hellogo/pkg/chess/move_test.go
package chess

import (
	"testing"
)

func TestValidMove(t *testing.T) {
	move := NewMove(0, 16)
	if move.From() != 0 {
		t.Errorf("expected from=0, got %d", move.From())
	}
	if move.To() != 16 {
		t.Errorf("expected to=16, got %d", move.To())
	}
}

func TestToString(t *testing.T) {
	move, err := NewMoveFromStrings("e2", "e4")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if move.String() != "e2e4" {
		t.Errorf("expected e2e4, got %s", move.String())
	}
}

func TestEquality(t *testing.T) {
	move1 := NewMove(0, 16)
	move2 := NewMove(0, 16)
	move3 := NewMove(16, 0)

	if !move1.Equals(move2) {
		t.Errorf("expected move1 == move2")
	}
	if move1.Equals(move3) {
		t.Errorf("expected move1 != move3")
	}
	// HashCode is not idiomatic in Go, so not tested.
}
